src/sparsebase/preprocess/preprocess.cc

```
@@ -415,6 +415,44 @@ IDType *RCMReorder<IDType, NNZType, ValueType>::GetReorderCSR(
   return Q;
 }
 
+#ifdef USE_AMD_ORDER
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include <amd.h>
+#ifdef __cplusplus
+}
+#endif
+
+template <typename IDType, typename NNZType, typename ValueType>
+AMDReorder<IDType, NNZType, ValueType>::AMDReorder(AMDReorderParams p){
+  this->params_ = std::make_unique<AMDReorderParams>(p);
+  this->RegisterFunction({format::CSR<IDType, NNZType, ValueType>::get_id_static()},
+                         AMDReorderCSR);
+}
+
+template <typename IDType, typename NNZType, typename ValueType>
+IDType* AMDReorder<IDType, NNZType, ValueType>::AMDReorderCSR(std::vector<format::Format*> formats, PreprocessParams* p){
+  auto csr = formats[0]->AsAbsolute<format::CSR<IDType, NNZType, ValueType>>();
+  auto csr_converted = csr->template Convert<format::CSR, long, long, ValueType>(false);
+  auto params = static_cast<AMDReorderParams*>(p);
+  long long n = csr->get_dimensions()[0];
+  long *xadj_long = csr_converted->get_row_ptr(), *adj_long = csr_converted->get_col();
+  long * i_order = new long[n];
+  double *Control = new double[AMD_CONTROL];
+  Control[AMD_DENSE] = params->dense;
+  Control[AMD_AGGRESSIVE] = params->aggressive ;
+  double *Info = nullptr; //new double[AMD_INFO]; // Auxiliary data
+  int status = amd_l_order(n, xadj_long, adj_long, i_order, Control, Info);
+  if (status != 0){
```

it is better to throw a meaningful error here.

---

also nice to extend our exceptions :).